File: src/main/java/com/pubnub/api/managers/PublishSequenceManager.java
package com.pubnub.api.managers;

public class PublishSequenceManager {


    private int maxSequence;
    private int nextSequence;

    public PublishSequenceManager(final int providedMaxSequence) {
        this.maxSequence = providedMaxSequence;
    }

    public final synchronized int getNextSequence() {
        if (maxSequence == nextSequence) {
            nextSequence = 1;
        } else {
            nextSequence += 1;
        }

        return nextSequence;
    }

}

File: src/main/kotlin/com/pubnub/api/managers/PublishSequenceManager.kt
package com.pubnub.api.managers

import java.util.concurrent.atomic.AtomicInteger

internal class PublishSequenceManager(private val maxSequence: Int) {

    private val atomicSeq = AtomicInteger(1)

    internal fun nextSequence(): Int = atomicSeq.getAndUpdate {
        if (maxSequence == it) {
            1
        } else {
            it + 1
        }
    }
}
